gpu/gpu.go

```diff
@@ -310,6 +339,39 @@ func GetGPUInfo() GpuInfoList {
 			}
 		}
 
+		// Vulkan
+		vHandles = initVulkanHandles()
+		for i := range vHandles.deviceCount {
+			if vHandles.vulkan != nil {
+				gpuInfo := VulkanGPUInfo{
+					GpuInfo: GpuInfo{
+						Library: "vulkan",
+					},
+					index: i,
+				}
+
+				C.vk_check_vram(*vHandles.vulkan, C.int(i), &memInfo)
```

Comment:
it could be nice to have a debugging log here printing the amount of memory detected.  (especially with iGPUs this number can be useful)

Reply:
Doesn't ollama do it already? When I was debugging I saw something like `Jun 15 20:25:32 rofl strace[403896]: time=2024-06-15T20:25:32.702+08:00 level=INFO source=types.go:102 msg="inference compute" id=0 library=vulkan compute=1.3 driver=1.3 name="Intel(R) Arc(tm) A770 Graphics (DG2)" total="15.9 GiB" available="14.3 GiB"
`

Reply:
I think you're right, but I don't that line exactly.   Looks like a CAP_PERFMON thing or I messed up compilation:

ggml_vulkan: Found 1 Vulkan devices:
Vulkan0: Intel(R) Iris(R) Plus Graphics (ICL GT2) | uma: 1 | fp16: 1 | warp size: 32

llama_new_context_with_model: Vulkan_Host  output buffer size =     0.14 MiB
llama_new_context_with_model:    Vulkan0 compute buffer size =   234.06 MiB
llama_new_context_with_model: Vulkan_Host compute buffer size =    24.01 MiB

**Vulkan.time=2024-06-16T13:20:27.582+02:00 level=DEBUG source=gpu.go:649 msg="Unable to load vulkan" library=/usr/lib64/libvulkan.so.1.3.279 /usr/lib64/libcap.so.2.69=error !BADKEY="performance monitoring is not allowed. Please enable CAP_PERFMON or run as root to use Vulkan."**

nvtop reveals iGPU being used as expected.

Reply:
Maybe run ollama as root? Or do `setcap cap_perfmon=+ep /path/to/ollama`

Reply:
thanks

setcap didn't work for some reason, I still get CAP_PERFMON errors. But running with sudo gives:

**time=2024-06-16T13:52:35.115+02:00 level=INFO source=gpu.go:355 msg="error looking up vulkan GPU memory" error="device is a CPU"
time=2024-06-16T13:52:35.130+02:00 level=INFO source=types.go:102 msg="inference compute" id=0 library=oneapi compute="" driver=0.0 name="Intel(R) Iris(R) Plus Graphics" total="0 B" available="0 B"
time=2024-06-16T13:52:35.130+02:00 level=INFO source=types.go:102 msg="inference compute" id=0 library=vulkan compute=1.3 driver=1.3 name="Intel(R) Iris(R) Plus Graphics (ICL GT2)" total="11.4 GiB" available="8.4 GiB"**

Reply:
Vulkan is reporting that the device is a CPU. If it's an iGPU it should've been detected.

You mentioned the performance was similar to when you were testing your branch. Are you sure you are not using CPU inference the entire time? Can you compare the performance against a CPU runner like cpu_avx?

Reply:
On a second read nevermind. It seems like everything is working as expected. Ollama detected two Vulkan devices, one is a CPU software implementation, which is skipped according to the error message, and the last line reports a Vulkan device that is recognized by ollama.

Reply:
Yes, that looks right.  There is also a lot of oneAPI junk in the logs that confuses me.  But it looks like Vulkan works as intended, but I have a CAP_PERFMON problem.

nvtop screenshot below:

![2024-06-16_14-33](https://github.com/ollama/ollama/assets/112642920/db6cda7b-0a3f-4b78-9613-e65737db3ad8)

I wonder why setcap does not work...  Could it be that one of the shared libraries (like libcap or libvulkan) needs a setcap instead of ollama binary?

Reply:
ok, the CAP_PERFMON is likely due to something off in my system.  It's trying to load the 32bit library for some reason:

**time=2024-06-16T14:58:06.326+02:00 level=DEBUG source=gpu.go:649 msg="Unable to load vulkan" library=/usr/lib/libvulkan.so.1.3.279 /usr/lib32/libcap.so.2.69=error !BADKEY="Unable to load /usr/lib32/libcap.so.2.69 library to query for Vulkan GPUs: /usr/lib32/libcap.so.2.69: wrong ELF class: ELFCLASS32**"

Reply:
Loading 32bit is expected, it's not related because it'll just skip it when it realizes it can't load it.